## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and clean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [390]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('/Users/kalafejzo/Downloads/P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


# Note - for now, it is okay to ignore the warning about mixed types. 

/var/folders/n6/bfj06ywj4n76z1b2g4_97fpr0000gn/T/ipykernel_2182/2958149657.py:14: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  contrib = pd.read_csv('/Users/kalafejzo/Downloads/P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)
/var/folders/n6/bfj06ywj4n76z1b2g4_97fpr0000gn/T/ipykernel_2182/2958149657.py:14: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('/Users/kalafejzo/Downloads/P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [391]:
contrib.shape

(1125659, 18)

- The shape of the data matches the expectation, with there being 1125659 rows and 18 columns. When I open the file in VS code this is the same number of rows and columns that I see in the data, indicating that the data's shape matches the expectation and the csv file loaded in correctly. Additionally, in the documentation, I am seeing 18 different columns as well.

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [392]:
contrib.columns

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')

- Yes, all of the columns included are in the documentation. When I open up the documentation I am seeing the same matching columns, a total of 18 of them.

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [393]:
contrib.head()

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- cand_id looks accurate and matches the expectation with a 'P' followed by an 8 digit numerical code. cmnd_nm also matches the expectation with the candidate's last name followed by a comma and then the first name, and middle name if they have one. The contbr_st matches expectation as well, indicated by a 2-letter code for the state, all of which are CA for the first five rows.

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [394]:
contrib["election_tp"].values

array(['P2016', 'P2016', 'P2016', ..., 'P2016', 'P2016', 'P2016'],
      dtype=object)

- The election_tp variable contains information on the year of the election and what type of election the contribution was made for. The values match the documentation, in which the code starts with one letter that indicates which type of election followed by a 4-digit number that indicates the election year.

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [395]:
contrib.dtypes

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- *The datatypes for the contrb_zip is an object, rather than an integer, likely due to there being a combination of NA values and integer values. contb_receipt_amt contains floats that can store up to 64 bits, contb_receipt_dt contains datetime64[ns] datatypes which is a timestamp for when the contribution was made. 

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [396]:
contrib.count()

cmte_id              1125659
cand_id              1125659
cand_nm              1125659
contbr_nm            1125659
contbr_city          1125633
contbr_st            1125659
contbr_zip           1125564
contbr_employer       967757
contbr_occupation    1115260
contb_receipt_amt    1125659
contb_receipt_dt     1125659
receipt_desc           15045
memo_cd               144268
memo_text             501148
form_tp              1125659
file_num             1125659
tran_id              1125659
election_tp          1124234
dtype: int64

- cmte_id, cand_id, cand_nm, contbr_st, contb_receipt_ant, contb_receipt_dt, form_tp, file_num, and tran_id have the most non-nulls as indicated by the count method which counts all the non-null values. Every value is non-null for these columns. Receipt_desc has only 1% of it's values as non-nulls, memo_cd only has 13%, and memo_text only has 45%. Since these columns all have more than 50% of the values as null, I would recommend dropping these columns. 

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [397]:
print(contrib.groupby('cand_nm')['cand_id'].nunique())
print(contrib[contrib['cand_nm'] == 'Sanders, Bernard']['cand_id'].nunique()) #Checks for Bernie Sanders


cand_nm
Bush, Jeb                    1
Carson, Benjamin S.          1
Christie, Christopher J.     1
Clinton, Hillary Rodham      1
Cruz, Rafael Edward 'Ted'    1
Fiorina, Carly               1
Gilmore, James S III         1
Graham, Lindsey O.           1
Huckabee, Mike               1
Jindal, Bobby                1
Johnson, Gary                1
Kasich, John R.              1
Lessig, Lawrence             1
McMullin, Evan               1
O'Malley, Martin Joseph      1
Pataki, George E.            1
Paul, Rand                   1
Perry, James R. (Rick)       1
Rubio, Marco                 1
Sanders, Bernard             1
Santorum, Richard J.         1
Stein, Jill                  1
Trump, Donald J.             1
Walker, Scott                1
Webb, James Henry Jr.        1
Name: cand_id, dtype: int64
1


- Throughout my dataframe, each candidate name only has one unique candidate id, indicating that they all match. I did this by grouping the candidate names and each one has only one unique id. I also double checked the Bernie Sanders name, and it returns 1 unique id. Therefore, we have no issues and I have checked the data quality of these two columns is good.

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [398]:
contrib[contrib['contbr_st'] != 'CA']['contbr_st'].shape[0] == 0

True

- I checked to see if there were any contributions whose contbr_st column was not equal to 'CA', meaning it would be from another state. I did so by using boolean notation to select for the non CA rows of data, and made sure the number of rows of data was equal to 0, and it outputted true. Therefore, there are no contributions with states outside of CA.

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [399]:
print(contrib['tran_id'].duplicated().value_counts()) #Counts the number of duplicated and non duplicated values 
dupes = contrib[contrib['tran_id'].duplicated(keep=False)] #Filters for the duplicated rows
print(dupes)
print(dupes.groupby('cand_nm')['tran_id'].count().sort_values(ascending=False)) #Checks if some candidates have more duplicates
print(dupes.groupby('contbr_nm')['tran_id'].count().sort_values(ascending=False)) #Checks if some contributors have more duplicates
print(dupes.groupby('election_tp')['tran_id'].count().sort_values(ascending=False)) #Checks if some certain elections have more duplicates
print(dupes.groupby('contb_receipt_dt')['tran_id'].count().sort_values(ascending=False)) #Checks if some dates have more duplicates
print(dupes.groupby('contb_receipt_amt')['tran_id'].count().sort_values(ascending=False)) #Checks if some contribution amounts have more duplicates



tran_id
False    1122205
True        3454
Name: count, dtype: int64
           cmte_id    cand_id                    cand_nm  \
57       C00575795  P00003392    Clinton, Hillary Rodham   
90       C00575795  P00003392    Clinton, Hillary Rodham   
91       C00575795  P00003392    Clinton, Hillary Rodham   
174      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
175      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
...            ...        ...                        ...   
1125589  C00573519  P60005915        Carson, Benjamin S.   
1125596  C00573519  P60005915        Carson, Benjamin S.   
1125597  C00573519  P60005915        Carson, Benjamin S.   
1125598  C00573519  P60005915        Carson, Benjamin S.   
1125614  C00573519  P60005915        Carson, Benjamin S.   

                     contbr_nm    contbr_city contbr_st     contbr_zip  \
57             SANMUKH, BHAKTA      SAN DIMAS        CA 917,733,736.00   
90             ROBERT, MONKMAN        FREMONT        CA 945,365

- There are a total of 3,545 duplicate entries. Patterns I see for duplicates include certain candidates, such as Hillary Clinton, Ted Cruz, Ben Carson and Marco Rubio have the majority of the duplications. Additionally, certain contributors have multiple duplicates and the primary election has a significantly higher number of duplicates than the general election.

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [400]:
print(contrib[contrib['contb_receipt_amt']<0].shape[0]) #Finds the number of negative donations
contrib[contrib['contb_receipt_amt']<0].head() #Shows a few rows of negative donations

11896


cmte_id    cand_id                    cand_nm           contbr_nm  \
19   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
23   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
81   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
190  C00574624  P60006111  Cruz, Rafael Edward 'Ted'    NOWELL, DIANA L.   
213  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  LICHTY, ANDREW MR.   

                contbr_city contbr_st     contbr_zip  \
19                    CHICO        CA 959,289,507.00   
23                    CHICO        CA 959,289,507.00   
81                    CHICO        CA 959,289,507.00   
190  RANCHO SANTA MARGARITA        CA 926,884,928.00   
213               SAN DIEGO        CA 921,096,720.00   

                        contbr_employer     contbr_occupation  \
19                        SELF EMPLOYED               RANCHER   
23                        SELF EMPLOYED               RANCHER   
81                        SELF EMPLOYED               RANCHER   
190  CAPISTRAND UNIFIED SCHOOL DISTRICT  LIBRARIAN TECHNICIAN   
213                       SELF EMPLOYED           REAL ESTATE   

     contb_receipt_amt contb_receipt_dt              receipt_desc memo_cd  \
19              -25.00       2016-04-29  REDESIGNATION TO GENERAL       X   
23             -150.00       2016-04-29  REDESIGNATION TO GENERAL       X   
81              -60.00       2016-04-14  REDESIGNATION TO GENERAL       X   
190            -100.00       2016-04-11  REDESIGNATION TO GENERAL       X   
213             -25.00       2016-04-30  REDESIGNATION TO GENERAL       X   

                    memo_text form_tp  file_num         tran_id election_tp  
19   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826482B       P2016  
23   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826483B       P2016  
81   REDESIGNATION TO GENERAL   SA17A   1077664   SA17A.1827494       P2016  
190  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1639830B       P2016  
213  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826888B       P2016

- There are a total of 11,896 negative donations. Negative donations mean that the contributor got refunded or recieved the amount of money back. This could be from a card or bank decline for example. The records, indicated by the columns "memo_text" and "receipt_desc" match the expectation of why a negative donation would happen as they all say "redesignation to general."

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [401]:
contrib[
    (contrib['contb_receipt_dt'] < "2014-01-01") | 
    (contrib['contb_receipt_dt'] > "2016-06-07")
]

contrib['contb_receipt_dt'].dtype

dtype('<M8[ns]')

- Yes. There are 451,272 rows that have dates outside of the primary period, which means we will likely have to filter out for only the 2016 primary contribution data. The dtype for 'contb_receipt_dt' is of the data tuype M8[ns], so the dates are well-formatted for our analysis.

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- Yes, we have the correct number of columns and rows.

**1l.2** Do the records contain data for the questions we want to answer?

- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
    Our records contain data to answer this question. We have the 'contbr_zip' column which gives us the zipcode. Along with that, the 'contb_receipt_amt' column gives us the dollar amount of all the contributions.
- What day(s) of the month do most people donate?
    Our records contain data to answer this question as well. We have the 'contb_receipt_dt' column that will allow us to find the date of the month and the 'contb_receipt_amt' column to find the day of the month people donate.

Additionally, our records contain the data to filter out for what our boss told us to filter for. The 'election_tp' will allow us to only select for primary 2016 data and the 'cand_nm' column will allow us to focus on Bernie Sanders.

**1l.3** What columns are important?

- The 'contbr_zip', the 'contb_receipt_amt', and 'contb_receipt_dt' columns are the primary most important columns to answer the questions asked. The 'election_tp' and 'cand_nm' columns are important for filtering.

**1l.4** What columns can be dropped?

- The 'cmte_id', 'cand_id', 'contbr_nm', 'contbr_city', 'receipt_desc', 'contbr_employer', 'contbr_occupation', 'memo_cd', 'memo_text', 'form_tp', 'file_num' columns can all be dropped from the data as we don't need these columns to answer any of our questions or filter out the data.

**1l.5** What are the data problems?

- We need to fix and remove the duplicates from the data. Additionally, we need to filter out for only primary election 2016 data and drop the rows that are outside of the time frame. We also need to account for the negative contributions when looking at our total donation amounts. Lastly, we need to address the missing values.

**1l.6** List any assumptions so far:

- Each row equals one contribution record
- Contribution dates are formatted correctly as datetime64[ns] objects with the year than month and then date
- Zip codes indicate where the person is located
- NA fields are ok to ignore
- Negative contribution amounts indicate a refund and can be removed from the donation total
- Each contribution was accurately reported with the right election type and candidate name/id

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [402]:
contrib = contrib[contrib['election_tp'] == 'P2016']
print(contrib.shape)
contrib

(810481, 18)


cmte_id    cand_id                  cand_nm  \
0        C00575795  P00003392  Clinton, Hillary Rodham   
1        C00575795  P00003392  Clinton, Hillary Rodham   
2        C00575795  P00003392  Clinton, Hillary Rodham   
3        C00577130  P60007168         Sanders, Bernard   
4        C00577130  P60007168         Sanders, Bernard   
...            ...        ...                      ...   
1125654  C00573519  P60005915      Carson, Benjamin S.   
1125655  C00573519  P60005915      Carson, Benjamin S.   
1125656  C00573519  P60005915      Carson, Benjamin S.   
1125657  C00573519  P60005915      Carson, Benjamin S.   
1125658  C00573519  P60005915      Carson, Benjamin S.   

                      contbr_nm    contbr_city contbr_st     contbr_zip  \
0                    AULL, ANNE       LARKSPUR        CA 949,391,913.00   
1             CARROLL, MARYJEAN        CAMBRIA        CA 934,284,638.00   
2              GANDARA, DESIREE        FONTANA        CA 923,371,507.00   
3                     LEE, ALAN      CAMARILLO        CA 930,111,214.00   
4              LEONELLI, ODETTE  REDONDO BEACH        CA 902,784,310.00   
...                         ...            ...       ...            ...   
1125654  HENDERSON, YVONNE MRS.  REDONDO BEACH        CA      902781750   
1125655   HENDRICKS, JIM C. MR.           ETNA        CA      960270030   
1125656   HENDRICKS, JIM C. MR.           ETNA        CA      960270030   
1125657   HENDRICKS, JIM C. MR.           ETNA        CA      960270030   
1125658   HENDRICKS, JIM C. MR.           ETNA        CA      960270030   

                   contbr_employer  contbr_occupation  contb_receipt_amt  \
0                              NaN            RETIRED              50.00   
1                              NaN            RETIRED             200.00   
2                              NaN            RETIRED               5.00   
3        AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER              40.00   
4         VERICOR ENTERPRISES INC.         PHARMACIST              35.00   
...                            ...                ...                ...   
1125654                    RETIRED            RETIRED             200.00   
1125655                    RETIRED            RETIRED              25.00   
1125656                    RETIRED            RETIRED              25.00   
1125657                    RETIRED            RETIRED              50.00   
1125658                    RETIRED            RETIRED              50.00   

        contb_receipt_dt receipt_desc memo_cd  \
0             2016-04-26          NaN       X   
1             2016-04-20          NaN       X   
2             2016-04-02          NaN       X   
3             2016-03-04          NaN     NaN   
4             2016-03-05          NaN     NaN   
...                  ...          ...     ...   
1125654       2016-01-29          NaN     NaN   
1125655       2016-01-06          NaN     NaN   
1125656       2016-01-12          NaN     NaN   
1125657       2016-01-21          NaN     NaN   
1125658       2016-01-29          NaN     NaN   

                                   memo_text form_tp  file_num       tran_id  \
0                     * HILLARY VICTORY FUND    SA18   1091718      C4768722   
1                     * HILLARY VICTORY FUND    SA18   1091718      C4747242   
2                     * HILLARY VICTORY FUND    SA18   1091718      C4666603   
3        * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   VPF7BKWA097   
4        * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   VPF7BKX3MB3   
...                                      ...     ...       ...           ...   
1125654                                  NaN   SA17A   1073637  SA17.1185074   
1125655                                  NaN   SA17A   1073637  SA17.1120355   
1125656                                  NaN   SA17A   1073637  SA17.1131185   
1125657                                  NaN   SA17A   1073637  SA17.1156676   
1125658                        

**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [ ]:
contrib = contrib[contrib['cand_nm'] == 'Sanders, Bernard']
print(contrib.shape)
contrib

(407171, 18)


cmte_id    cand_id           cand_nm            contbr_nm  \
3        C00577130  P60007168  Sanders, Bernard            LEE, ALAN   
4        C00577130  P60007168  Sanders, Bernard     LEONELLI, ODETTE   
5        C00577130  P60007168  Sanders, Bernard     LEONELLI, ODETTE   
6        C00577130  P60007168  Sanders, Bernard       LEOPARD, PATTI   
8        C00577130  P60007168  Sanders, Bernard         LEPKE, KELLY   
...            ...        ...               ...                  ...   
1121696  C00577130  P60007168  Sanders, Bernard       SIMONE, ARLENE   
1121698  C00577130  P60007168  Sanders, Bernard       SIMONE, ARLENE   
1121703  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   
1121704  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   
1121705  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   

           contbr_city contbr_st     contbr_zip            contbr_employer  \
3            CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4        REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   
5        REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   
6                VISTA        CA 920,842,849.00  ONSITE ENERGY CORPORATION   
8          WESTMINSTER        CA 926,833,846.00                       NONE   
...                ...       ...            ...                        ...   
1121696         ENCINO        CA      914361655     SELF EMPLOYED-EMPLOYED   
1121698         ENCINO        CA      914361655     SELF EMPLOYED-EMPLOYED   
1121703      FULLERTON        CA      928313414        CITY OF LOS ANGELES   
1121704      FULLERTON        CA      928313414        CITY OF LOS ANGELES   
1121705      FULLERTON        CA      928313414        CITY OF LOS ANGELES   

            contbr_occupation  contb_receipt_amt contb_receipt_dt  \
3           SOFTWARE ENGINEER              40.00       2016-03-04   
4                  PHARMACIST              35.00       2016-03-05   
5                  PHARMACIST             100.00       2016-03-06   
6             PROJECT MANAGER              25.00       2016-03-04   
8                NOT EMPLOYED              10.00       2016-03-05   
...                       ...                ...              ...   
1121696          ENTERTASINER              50.00       2016-04-23   
1121698          ENTERTASINER             100.00       2016-04-27   
1121703  MANAGEMENT ASSISTANT              15.00       2016-04-15   
1121704  MANAGEMENT ASSISTANT              15.00       2016-04-23   
1121705  MANAGEMENT ASSISTANT              15.00       2016-04-25   

        receipt_desc memo_cd                            memo_text form_tp  \
3                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
4                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
5                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
6                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
8                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
...              ...     ...                                  ...     ...   
1121696          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121698          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121703          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121704          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121705          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   

         file_num      tran_id election_tp  
3         1077404  VPF7BKWA097       P2016  
4         1077404  VPF7BKX3MB3       P2016  
5         1077404  VPF7BKYBXV4       P2016  
6         1077404  VPF7BKW04C1       P2016  
8         1077404  VPF7BKX3H59       P2016  
...           ...          ...         ...  
1121696   1079445  VPF7BP5TFS6       P2016  
1121698   1079445  VPF7BP86GH6       P2016  
1121703   1079445  VPF7BNXXDY5       P2016  
11217

**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [404]:
contrib['contbr_zip'] = contrib['contbr_zip'].astype(str)
contrib['five_dig_zip'] = contrib['contbr_zip'].str[:5].str.replace(".", "")
contrib = contrib.drop('contbr_zip', axis=1)
contrib['five_dig_zip'] = contrib['five_dig_zip'].astype(int)
contrib = contrib[
    (contrib['five_dig_zip'] > 90001) & (contrib['five_dig_zip'] < 96162)
    ]

/var/folders/n6/bfj06ywj4n76z1b2g4_97fpr0000gn/T/ipykernel_2182/1287872328.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  contrib['contbr_zip'] = contrib['contbr_zip'].astype(str)
/var/folders/n6/bfj06ywj4n76z1b2g4_97fpr0000gn/T/ipykernel_2182/1287872328.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  contrib['five_dig_zip'] = contrib['contbr_zip'].str[:5].str.replace(".", "")


**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [405]:
contrib = contrib[contrib['contb_receipt_amt'] > 0]
print(contrib.shape)
contrib

(404011, 18)


cmte_id    cand_id           cand_nm            contbr_nm  \
3        C00577130  P60007168  Sanders, Bernard            LEE, ALAN   
4        C00577130  P60007168  Sanders, Bernard     LEONELLI, ODETTE   
5        C00577130  P60007168  Sanders, Bernard     LEONELLI, ODETTE   
6        C00577130  P60007168  Sanders, Bernard       LEOPARD, PATTI   
8        C00577130  P60007168  Sanders, Bernard         LEPKE, KELLY   
...            ...        ...               ...                  ...   
1121696  C00577130  P60007168  Sanders, Bernard       SIMONE, ARLENE   
1121698  C00577130  P60007168  Sanders, Bernard       SIMONE, ARLENE   
1121703  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   
1121704  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   
1121705  C00577130  P60007168  Sanders, Bernard  ROBINSON, KATHARINE   

           contbr_city contbr_st            contbr_employer  \
3            CAMARILLO        CA  AT&T GOVERNMENT SOLUTIONS   
4        REDONDO BEACH        CA   VERICOR ENTERPRISES INC.   
5        REDONDO BEACH        CA   VERICOR ENTERPRISES INC.   
6                VISTA        CA  ONSITE ENERGY CORPORATION   
8          WESTMINSTER        CA                       NONE   
...                ...       ...                        ...   
1121696         ENCINO        CA     SELF EMPLOYED-EMPLOYED   
1121698         ENCINO        CA     SELF EMPLOYED-EMPLOYED   
1121703      FULLERTON        CA        CITY OF LOS ANGELES   
1121704      FULLERTON        CA        CITY OF LOS ANGELES   
1121705      FULLERTON        CA        CITY OF LOS ANGELES   

            contbr_occupation  contb_receipt_amt contb_receipt_dt  \
3           SOFTWARE ENGINEER              40.00       2016-03-04   
4                  PHARMACIST              35.00       2016-03-05   
5                  PHARMACIST             100.00       2016-03-06   
6             PROJECT MANAGER              25.00       2016-03-04   
8                NOT EMPLOYED              10.00       2016-03-05   
...                       ...                ...              ...   
1121696          ENTERTASINER              50.00       2016-04-23   
1121698          ENTERTASINER             100.00       2016-04-27   
1121703  MANAGEMENT ASSISTANT              15.00       2016-04-15   
1121704  MANAGEMENT ASSISTANT              15.00       2016-04-23   
1121705  MANAGEMENT ASSISTANT              15.00       2016-04-25   

        receipt_desc memo_cd                            memo_text form_tp  \
3                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
4                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
5                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
6                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
8                NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
...              ...     ...                                  ...     ...   
1121696          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121698          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121703          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121704          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
1121705          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   

         file_num      tran_id election_tp  five_dig_zip  
3         1077404  VPF7BKWA097       P2016         93011  
4         1077404  VPF7BKX3MB3       P2016         90278  
5         1077404  VPF7BKYBXV4       P2016         90278  
6         1077404  VPF7BKW04C1       P2016         92084  
8         1077404  VPF7BKX3H59       P2016         92683  
...           ...          ...         ...           ...  
1121696   1079445  VPF7BP5TFS6       P2016         91436  
1121698   1079445  VPF7BP86GH6       P2016         91436  
1121703   1079445  VPF7BNXXDY5       P2016         92831  
1121704   1079445  VPF7BP64DH7       P2016   

**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [406]:
contrib = contrib.drop(
    ['cmte_id', 'cand_id', 'contbr_nm', 'contbr_city', 'receipt_desc', 'contbr_employer', 'contbr_occupation', 'memo_cd', 'memo_text', 'form_tp', 'file_num'], 
    axis=1)
print(contrib.shape)
contrib

(404011, 7)


cand_nm contbr_st  contb_receipt_amt contb_receipt_dt  \
3        Sanders, Bernard        CA              40.00       2016-03-04   
4        Sanders, Bernard        CA              35.00       2016-03-05   
5        Sanders, Bernard        CA             100.00       2016-03-06   
6        Sanders, Bernard        CA              25.00       2016-03-04   
8        Sanders, Bernard        CA              10.00       2016-03-05   
...                   ...       ...                ...              ...   
1121696  Sanders, Bernard        CA              50.00       2016-04-23   
1121698  Sanders, Bernard        CA             100.00       2016-04-27   
1121703  Sanders, Bernard        CA              15.00       2016-04-15   
1121704  Sanders, Bernard        CA              15.00       2016-04-23   
1121705  Sanders, Bernard        CA              15.00       2016-04-25   

             tran_id election_tp  five_dig_zip  
3        VPF7BKWA097       P2016         93011  
4        VPF7BKX3MB3       P2016         90278  
5        VPF7BKYBXV4       P2016         90278  
6        VPF7BKW04C1       P2016         92084  
8        VPF7BKX3H59       P2016         92683  
...              ...         ...           ...  
1121696  VPF7BP5TFS6       P2016         91436  
1121698  VPF7BP86GH6       P2016         91436  
1121703  VPF7BNXXDY5       P2016         92831  
1121704  VPF7BP64DH7       P2016         92831  
1121705  VPF7BP6KBS4       P2016         92831  

[404011 rows x 7 columns]

- I dropped the following columns: cmte_id', 'cand_id', 'contbr_nm', 'contbr_city', 'receipt_desc', 'contbr_employer', 'contbr_occupation', 'memo_cd', 'memo_text', 'form_tp', 'file_num' because none of them are needed in answering the questions or filtering out the data.

**2f.** List any assumptions that you made up to this point:

- That every cand_nm was accurately reported as voting or Bernie Sanders
- election_tp was accurately reported as P2016 if it was a contribution for the primary 2016 election
- That it is more accurate to just completely drop the negative donation amounts rather than keeping them and subtracting them from the positive amounts.
- Zipcodes rather than the state columns are an accurate way to determine whether a contribution was made in CA

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [413]:
dollar_stats = contrib.groupby('five_dig_zip')['contb_receipt_amt'].sum()
cont_stats = contrib.groupby('five_dig_zip').size()
print(dollar_stats.sort_values(ascending=False))
print(cont_stats.sort_values(ascending=False))

five_dig_zip
94110   294,061.13
94114   208,602.37
94117   201,545.43
94611   150,232.12
95060   145,414.63
           ...    
95233        15.00
95703        15.00
95701        15.00
93483        10.00
92365        10.00
Name: contb_receipt_amt, Length: 1832, dtype: float64
five_dig_zip
94110    4046
94114    3043
94117    2884
95060    2849
94611    2516
         ... 
93530       1
92616       1
92079       1
93590       1
93190       1
Length: 1832, dtype: int64


- The 94110 zipcode had both the highest dollar amount and the highest number of contributers, followed by 94114 and 94117.

**3b.** What day(s) of the month do most people donate? (10 points)

In [416]:
date_data = contrib.copy()
date_data['day_of_week'] = date_data['contb_receipt_dt'].astype(str).str[-2::]
date_counts = date_data.groupby('day_of_week').size()
print(date_counts.sort_values(ascending=False))

day_of_week
29    23504
30    22897
31    19484
14    17791
08    16772
28    16215
09    16109
16    14179
13    13797
10    13366
27    13079
17    12886
26    12537
23    12154
15    12024
04    11946
01    11938
18    11935
06    11615
20    11214
02    11065
21    10712
05    10532
07    10474
11    10409
12    10262
25     9893
03     9466
19     8952
22     8583
24     8221
dtype: int64


- People typically donate towards the end of the month, with the 29th being the highest day, followed by 30th and then the 14th. 

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k